### PR TITLE
[FLINK-4068] [table] Move constant computations out of code-generated

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

The origin PR is #2102. This PR makes `ReduceExpressionsRules` take effect. This rule can reduce constant expressions and replacing them with the corresponding constant.

And add some tests.

### CODE REVIEW

**Comment by twalthr** [comment]
I will shepherd this PR.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/BatchTableEnvironmentTest.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+package org.apache.flink.api.table
+
+import org.apache.flink.api.scala.ExecutionEnvironment
+import org.apache.flink.api.scala.table._
+import org.apache.flink.api.scala.util.CollectionDataSets
+import org.junit.Assert._
+import org.junit.Test
+
+
+class BatchTableEnvironmentTest {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/table/BatchTableEnvironmentTest.scala]: Could you add some additional expressions for better testing? Boolean expressions, something with `NULL`, `CASE`, strings and date/time types.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/BatchTableEnvironment.scala [hunk 2]
```diff
@@ -228,19 +229,13 @@ abstract class BatchTableEnvironment(
   }
 
   /**
-    * Translates a [[Table]] into a [[DataSet]].
-    *
-    * The transformation involves optimizing the relational expression tree as defined by
+    * Generates the optimized [[RelNode]] tree from the relational expression tree as defined by
     * Table API calls and / or SQL queries and generating corresponding [[DataSet]] operators.
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/BatchTableEnvironment.scala]: The method does not generate a `DataSet` operators. Can you also add that it converts to DataSet/DataStream convertion?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkRelBuilder.scala [hunk 3]
```diff
@@ -38,7 +38,12 @@ class FlinkRelBuilder(
     cluster,
     relOptSchema) {
 
-  def getPlanner: RelOptPlanner = cluster.getPlanner
+  def getPlanner: RelOptPlanner = {
+    val planner = cluster.getPlanner
+    // set the executor to evaluate constant expressions
+    planner.setExecutor(new RexExecutorImpl(Schemas.createDataContext(null)))
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkRelBuilder.scala]: good idea!

**Comment by wuchong** [comment]
Hi @twalthr , I have updated this PR , could you have a look again ?

**Comment by twalthr** [comment]
Thanks for updating the PR @wuchong. I found a little bug, but I will fix it myself and merge this.

**Comment by twalthr** [comment]
@wuchong it seems that the constant expression reduction does not work in the Table API . Could you convert your SQL tests also to additional Table tests and have a look at it? The SQL part seems to work fine.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkPlannerImpl.scala [hunk 4]
```diff
@@ -60,6 +60,7 @@ class FlinkPlannerImpl(
   var root: RelRoot = _
 
   private def ready() {
+    planner.setExecutor(config.getExecutor)
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkPlannerImpl.scala]: I would not set the planner here as this class is only used by the SQL API. `FlinkRelBuilder.create()` is a better place.

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/FlinkPlannerImpl.scala]: Yes you are right ! This is the reason why Table API not work.

**Comment by wuchong** [comment]
@twalthr  I add some Table API tests but skip some tests doesn't supported in Table API, such as `NULLIF`, `IN`, `EXTRACT`

**Comment by twalthr** [comment]
@wuchong I heavily tested this PR again as it could introduce difficult bugs. And I also found one issue that needs to be discussed. If you have an expression like `CAST(TRUE AS VARCHAR) || 'Test'` the result will be `TTest` not `trueTest`. So the expression reduction has a different string representation of boolean literals. The question is, do we wanna adapt our code, can this be configured, or is Calcite doing something wrong? What do you think @fhueske?

**Pull request closed without merging**